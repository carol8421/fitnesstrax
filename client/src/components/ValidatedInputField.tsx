import { Option } from "ld-ambiguity"
import React from "react"

import * as csn from "../classnames"
import InputField from "./InputField"

//import { maybe, isSomething } from "../common"

/* Required parameters:
 *
 * value: the data value to be rendered
 * render: (value) => str
 * parser: (str) => maybe value
 *
 * Optional parameter:
 *
 * onChange: (value) => action to take when a valid value is in the field
 */

export interface Props<A> {
  value: Option<A>
  placeholder: string
  render: (_: A) => string
  parse: (_: string) => Option<A>
  onChange: ((_: A) => void)
}

interface State<A> {
  text: Option<string>
  value: Option<A>
}

export class ValidatedInputField<A> extends React.Component<
  Props<A>,
  State<A>
> {
  state: State<A>

  constructor(props: Props<A>) {
    super(props)

    this.state = {
      text: props.value.map(props.render),
      value: props.value,
    }
  }

  update(value: string) {
    const { parse, onChange } = this.props
    const res = parse(value)
    this.setState({ text: Option.Some(value), value: res })
    res.map(v => onChange(v))
  }

  render() {
    const { placeholder } = this.props
    const { value } = this.state
    const valid = value
      .map(_ => ({
        valid: true,
      }))
      .or({ valid: false })
    return (
      <span className={csn.classnames(valid)}>
        <InputField
          classNames={valid}
          value={this.state.text}
          onChange={value => this.update(value)}
          placeholder={placeholder}
        />
      </span>
    )
  }
}

export default ValidatedInputField
