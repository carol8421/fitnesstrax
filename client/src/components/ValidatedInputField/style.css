input {
  margin: 0.25em;
}

input.invalid {
  border: 2px solid rgb(255, 96, 96);
  padding: 2px;
  border-radius: 2px;
}
