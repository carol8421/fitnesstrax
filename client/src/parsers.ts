import { Option, Result, sequenceResult } from "ld-ambiguity"
import _ from "lodash/fp"
import { Duration } from "luxon"

export const parseNumber = (str: string): Result<Option<number>, string> => {
  if (str === "") return Result.Ok(Option.None())
  const pattern = /^[0-9]+(.[0-9]+)?$/

  if (pattern.test(str)) {
    const res = Option.fromNaN(parseFloat(str))
    if (res.isSome()) {
      return Result.Ok(res)
    } else {
      return Result.Err("parse failed")
    }
  } else {
    return Result.Err("parse failed")
  }
}

export const parseTime = (
  str: string,
): Result<
  Option<{ hours: number; minutes: number; seconds: number }>,
  string
> => {
  if (str === "") return Result.Ok(Option.None())

  const lst: Result<Array<Option<number>>, string> = sequenceResult(
    _.map(
      (v: string): Result<Option<number>, string> =>
        parseNumber(v).andThen(
          v => (v.isSome() ? Result.Ok(v) : Result.Err("invalid time string")),
        ),
    )(str.split(":")),
  )
  if (lst.isOk()) {
    const lst_ = lst.unwrap()
    if (lst_.length === 2) {
      return Result.Ok(
        Option.Some({
          hours: lst_[0].unwrap(),
          minutes: lst_[1].unwrap(),
          seconds: 0,
        }),
      )
    } else if (lst_.length === 3) {
      return Result.Ok(
        Option.Some({
          hours: lst_[0].unwrap(),
          minutes: lst_[1].unwrap(),
          seconds: lst_[2].unwrap(),
        }),
      )
    }
    return Result.Ok(Option.None())
  } else {
    return Result.Err("invalid time string")
  }
}

export const parseDuration = (
  str: string,
): Result<Option<Duration>, string> => {
  if (str === "") return Result.Ok(Option.None())
  const lst: Result<Array<Option<number>>, string> = sequenceResult(
    _.map(
      (v: string): Result<Option<number>, string> =>
        parseNumber(v).andThen(
          v =>
            v.isSome() ? Result.Ok(v) : Result.Err("invalid duration string"),
        ),
    )(str.split(":")),
  )

  if (lst.isOk()) {
    const lst_ = lst.unwrap()
    if (lst_.length === 1) {
      return Result.Ok(
        Option.Some(Duration.fromObject({ seconds: lst_[0].unwrap() })),
      )
    } else if (lst_.length === 2) {
      return Result.Ok(
        Option.Some(
          Duration.fromObject({
            minutes: lst_[0].unwrap(),
            seconds: lst_[1].unwrap(),
          }),
        ),
      )
    } else if (lst_.length === 3) {
      return Result.Ok(
        Option.Some(
          Duration.fromObject({
            hours: lst_[0].unwrap(),
            minutes: lst_[1].unwrap(),
            seconds: lst_[2].unwrap(),
          }),
        ),
      )
    } else {
      return Result.Err("invalid duration string")
    }
  } else {
    return Result.Err("invalid duration string")
  }
}

/*
export const parseDateTimeTz = (inp: string): Result<Option<DateTimeTz>, string> => {
  if (inp === "") return Result.Ok(Option.None())
  const res = parseTime(inp).map(
                t =>
                  new DateTimeTz(
                    DateTime.fromObject({
                      year: date.year,
                      month: date.month,
                      day: date.day,
                      hour: t.hours,
                      minute: t.minutes,
                      second: t.seconds,
                      zone: prefs.timezone,
                    }),
                  ),
              )
  if (res.isSome()) {
  }
  return Result.Err("
}
   */
