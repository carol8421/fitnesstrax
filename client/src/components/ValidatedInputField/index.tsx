import { Option, Result } from "ld-ambiguity"
import React from "react"

import * as csn from "../../classnames"
import InputField from "../InputField"
import "./style.css"

/* Required parameters:
 *
 * value: the data value to be rendered
 * render: (value) => str
 * parser: (str) => maybe value
 *
 * Optional parameter:
 *
 * onChange: (value) => action to take when a valid value is in the field
 */

export interface Props<A> {
  value: Option<A>
  placeholder: string
  render: (_: A) => string
  parse: (_: string) => Result<Option<A>, string>
  onChange: ((_: Option<A>) => void)
}

interface State<A> {
  text: string
  value: Option<A>
  valid: boolean
}

export class ValidatedInputField<A> extends React.Component<
  Props<A>,
  State<A>
> {
  state: State<A>

  constructor(props: Props<A>) {
    super(props)

    this.state = {
      text: props.value.map(props.render).or(""),
      value: props.value,
      valid: true,
    }
  }

  update(value: string) {
    const { parse, onChange } = this.props
    const res = parse(value)
    if (res.isOk()) {
      this.setState({ text: value, value: res.unwrap(), valid: true })
      res.map(v => onChange(v))
    } else {
      this.setState({ text: value, value: Option.None(), valid: false })
    }
  }

  render() {
    const { placeholder } = this.props
    const { value, valid } = this.state
    const validityClass = valid
      ? { valid: true, invalid: false }
      : { valid: false, invalid: true }
    return (
      <div className={csn.classnames(validityClass)}>
        <InputField
          classNames={validityClass}
          value={this.state.text}
          onChange={value => this.update(value)}
          placeholder={placeholder}
        />
      </div>
    )
  }
}

export default ValidatedInputField
