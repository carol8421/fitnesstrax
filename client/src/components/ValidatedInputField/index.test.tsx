import { mount } from "enzyme"
import { Option } from "ld-ambiguity"
import React from "react"

import { parseNumber } from "../../parsers"
import ValidatedInputField from "./index"

describe("ValidatedInputField", () => {
  it("renders", () => {
    const onChange = jest.fn()
    const wrapper = mount(
      <ValidatedInputField
        value={Option.Some(15)}
        placeholder="validated input field"
        render={(val: number): string => val.toString()}
        parse={parseNumber}
        onChange={onChange}
      />,
    )
    expect(wrapper).toMatchSnapshot()
    expect(onChange).not.toHaveBeenCalled()
    expect(wrapper.find("input").prop("value")).toEqual("15")
  })

  it("handles valid updates", () => {
    const onChange = jest.fn()
    const wrapper = mount(
      <ValidatedInputField
        value={Option.Some(15)}
        placeholder="validated input field"
        render={(val: number): string => val.toString()}
        parse={parseNumber}
        onChange={onChange}
      />,
    )
    wrapper.find("input").simulate("change", { target: { value: "16" } })
    expect(wrapper).toMatchSnapshot()
    expect(onChange).toHaveBeenCalledWith(Option.Some(16))
    expect(wrapper.find("input").prop("value")).toEqual("16")
    expect(wrapper.find("input").prop("className")).toEqual("valid")
  })

  it("handles invalid updates", () => {
    const onChange = jest.fn()
    const wrapper = mount(
      <ValidatedInputField
        value={Option.Some(15)}
        placeholder="validated input field"
        render={(val: number): string => val.toString()}
        parse={parseNumber}
        onChange={onChange}
      />,
    )
    wrapper.find("input").simulate("change", { target: { value: "abcdefg" } })
    expect(wrapper).toMatchSnapshot()
    expect(onChange).not.toHaveBeenCalled()
    expect(wrapper.find("input").prop("value")).toEqual("abcdefg")
    expect(wrapper.find("input").prop("className")).toEqual("invalid")
  })
})
