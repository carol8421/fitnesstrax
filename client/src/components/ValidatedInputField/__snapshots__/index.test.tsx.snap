// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatedInputField handles invalid updates 1`] = `
<ValidatedInputField
  onChange={[MockFunction]}
  parse={[Function]}
  placeholder="validated input field"
  render={[Function]}
  value={
    t {
      "val_": 15,
    }
  }
>
  <div
    className="invalid"
  >
    <InputField
      classNames={
        Object {
          "invalid": true,
          "valid": false,
        }
      }
      onChange={[Function]}
      placeholder="validated input field"
      value="abcdefg"
    >
      <input
        className="invalid"
        onChange={[Function]}
        placeholder="validated input field"
        type="text"
        value="abcdefg"
      />
    </InputField>
  </div>
</ValidatedInputField>
`;

exports[`ValidatedInputField handles valid updates 1`] = `
<ValidatedInputField
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          t {
            "val_": 16,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  parse={[Function]}
  placeholder="validated input field"
  render={[Function]}
  value={
    t {
      "val_": 15,
    }
  }
>
  <div
    className="valid"
  >
    <InputField
      classNames={
        Object {
          "invalid": false,
          "valid": true,
        }
      }
      onChange={[Function]}
      placeholder="validated input field"
      value="16"
    >
      <input
        className="valid"
        onChange={[Function]}
        placeholder="validated input field"
        type="text"
        value="16"
      />
    </InputField>
  </div>
</ValidatedInputField>
`;

exports[`ValidatedInputField renders 1`] = `
<ValidatedInputField
  onChange={[MockFunction]}
  parse={[Function]}
  placeholder="validated input field"
  render={[Function]}
  value={
    t {
      "val_": 15,
    }
  }
>
  <div
    className="valid"
  >
    <InputField
      classNames={
        Object {
          "invalid": false,
          "valid": true,
        }
      }
      onChange={[Function]}
      placeholder="validated input field"
      value="15"
    >
      <input
        className="valid"
        onChange={[Function]}
        placeholder="validated input field"
        type="text"
        value="15"
      />
    </InputField>
  </div>
</ValidatedInputField>
`;
